tokio-timer/src/timer/entry.rs

```diff
@@ -0,0 +1,559 @@
+                return;
+            }
+
+            let next = ERROR;
+
+            let actual = self.state.compare_and_swap(curr, next, SeqCst);
+
+            if curr == actual {
+                break;
+            }
+
+            curr = actual;
+        }
+
+        self.task.notify();
+    }
+
+    pub fn cancel(entry: &Arc<Entry>) {
+        let state = entry.state.fetch_or(ELAPSED, SeqCst);
+
+        if is_elapsed(state) {
+            // Nothing more to do
+            return;
+        }
+
+        let inner = match entry.inner.upgrade() {
+            Some(inner) => inner,
+            None => return,
+        };
+
+        let _ = inner.queue(entry);
+    }
+
+    pub fn poll_elapsed(&self) -> Poll<(), Error> {
+        use futures::Async::NotReady;
+
+        let mut curr = self.state.load(SeqCst);
+
+        if is_elapsed(curr) {
+            if curr == ERROR {
+                return Err(Error::shutdown());
+            } else {
+                return Ok(().into());
+            }
+        }
+
+        self.task.register();
+
+        curr = self.state.load(SeqCst).into();
+
+        if is_elapsed(curr) {
+            if curr == ERROR {
+                return Err(Error::shutdown());
+            } else {
+                return Ok(().into());
+            }
+        }
+
+        Ok(NotReady)
+    }
+
+    pub fn reset(entry: &Arc<Entry>, deadline: Instant) {
+        let inner = match entry.inner.upgrade() {
+            Some(inner) => inner,
+            None => return,
+        };
+
+        let when = inner.normalize_deadline(deadline);
+        let elapsed = inner.elapsed();
+
+        let mut curr = entry.state.load(SeqCst);
+        let mut notify;
+
+        loop {
+            if curr == ERROR {
+                // Registration in error state, nothing further to do.
+                return;
+            } else if curr == when {
+                // State already set to notify at `deadline`.
+                return;
+            }
+
+            let next;
+
+            if when <= elapsed {
+                next = ELAPSED;
+                notify = !is_elapsed(curr);
+            } else {
+                next = when;
+                notify = true;
+            }
+
+            let actual = entry.state.compare_and_swap(
+                curr, next, SeqCst);
+
+            if curr == actual {
+                break;
+            }
+
+            curr = actual;
+        }
+
+        if notify {
+            let _ = inner.queue(entry);
+        }
+    }
+}
+
+fn is_elapsed(state: u64) -> bool {
+    state & ELAPSED == ELAPSED
+}
+
+impl Drop for Entry {
+    fn drop(&mut self) {
+        if !self.counted {
+            return;
+        }
+
+        let inner = match self.inner.upgrade() {
+            Some(inner) => inner,
+            None => return,
+        };
+
+        inner.decrement();
+    }
+}
+
+unsafe impl Send for Entry {}
+unsafe impl Sync for Entry {}
+
+// ===== impl Stack =====
+
+impl Stack {
```

Comment:
Are you sure you want to implement stack from scratch? I am sure there is a library solution for your case.

Reply:
I am not aware of any crate that provides an intrusive doubly linked list based stack.

Reply:
There is https://docs.rs/crossbeam/0.3.2/crossbeam/sync/struct.TreiberStack.html. I'm not sure about its status in the new v0.4 however.

Reply:
I've seen this, but it is not intrusive and this case doesn't call for a treiber stack :)

Reply:
https://docs.rs/intrusive-collections/0.7.0/intrusive_collections/linked_list/struct.LinkedList.html? :)

Reply:
@hawkw At a glance, it could possibly work (though the API was not *immediately* obvious how to use).

That said, the impl has a tail pointer, which would result in doubling the size of the wheel. The timer does not require a tail pointer, only that nodes can remove themselves.

Reply:
I think the _sui generis_ implementation is probably best here, just felt like pointing out another option :)